TGUI for RCD , Radial menu for RLD & Silo Link for RLD & Plumbing RCD

SyncIt21 opened the pull request:
**About the pull request**
- TGUI window for RCD which replaces all its radial menus
- Radial menu for RLD which is more intuitive than clicking & checking the logs to see if you got the right mode
- Construction consoles both the aux & centcom also had their switch mode , door & airlock actions replaced with the same RCD TGUI interface as above with one action click
- RLD & Plumbing RCD can now benefit from silo link .

The whole demo is actually  67 MB & i had to compress it like 3 times cause github wont allow files larger than 10mb in PRs. So yeah dont expect 1080p

https://user-images.githubusercontent.com/110812394/205930334-f76b2d98-d83c-4fa9-9229-6136ad6ba0e3.mp4


**Why its goods for the game**
For the same reasons as #71569, all of us can agree that the radial menus are absolutely abused for the RCD with one menu leading to another menu and finally leading you into depression. That mess just has to go. Not to mention with the airlock electronics interface embedded in the window we don't have to open a separate TGUI window just for that.

Also i reused that interface inside the construction consoles thus allowing editing all aspects of the RCD with just one action button rather than having multiple buttons for build mode , glass type , airlock mode whatever.

Having a radial menu for the RLD is also more intuitive rather than just clicking & checking the logs to see if you got the correct mode The icons for light and glowsticks gets coloured based on your selected colour thus making your choice more obvious

Allowing the RLD & Plumbing to have Silo link makes sense as  `/obj/item/construction` has all the code required for installing silo links but there was no visual interface to enable these links. Now that's available for both RLD & plumbing RCD.
Don't want to hear anyone complain how this could be a power creep. Creating lot of lights , tubes , synthesizers etc aint gonna harm anyone. If everyone agreed that it's OK for the RCD to create lots of walls , doors etc  via the silo link then this shouldn't be a problem.

**Code Improvements**
1. Lot of `toggle(), change(), set()` procs have all been removed and all their functionality is now in the `ui_act()` function. These procs were very inefficient as they created a list of choices, loading icons for the radial menus every time they were invoked. That & the endless switch & if statements are just poorly written. 
Now there is one big global static list for all RCD choices & RLD choices as well

2. Plumbing RCD & the Normal RCD sprite files only load the necessary icon_states from each icon file thus reducing the final asset files size which equals a lot of good stuff

**Change Log**
:cl:
add: TGUI for RCD Just like Rapid Pipe Dispenser
add: Radial menu for RLD
code: silo link upgrades for both RLD & Plumbing RCD
del: Switch Mode, Window Glass, Airlock action icons for construction console
refactor: FURNISHING , FRAMES & SIMPLE_CIRCUITS are banned upgrades for RLD & Plumbing RCD
/:cl:

SyncIt21 committed:
RCD.dm lot of cool stuff

1. TGUI for RCD & Radial for RLD
2. Ore silo link for rld & plumbing RCD
3. Lot of code removal & optimizations

SyncIt21 committed:
Merge branch 'tgstation:master' into RCD-TGUI

SyncIt21 committed:
Slight tweaks after merging master

SyncIt21 committed:
Merge branch 'master' into RCD-TGUI

SyncIt21 commented:
I dont know to write proper changelogs. If anyone has a better suggestion put it as a comment & ill copy paste it.

SyncIt21 commented:
> You currently have a negative Fix/Feature pull request delta of -7. Maintainers may close this PR at will. Fixing issues or improving the codebase will improve this score.

Fuck u github

Fikou commented:
its refactor not refractor

SyncIt21 commented:
> its refactor not refractor

Thank u

SyncIt21 committed:
Call Parent Proc

SyncIt21 committed:
Merge branch 'RCD-TGUI' of https://github.com/SyncIt21/The-TG-Station-Fork into RCD-TGUI

SyncIt21 committed:
Typecast to list

SyncIt21 commented:
Bad icon operation? what am i supposed to use if not Blend() function? or is that even the source of the error here?

SyncIt21 committed:
Maybe its the single quotes?

SyncIt21 commented:
So it was the single quotes!! OOOFFF

MTandi commented:
Nice, but I wish it used space a bit better and needed less clicking. Maybe half the size of icons and use less tabs?

Ryusenshu commented:
Not a fan of this, seems WAY slower than before

SyncIt21 commented:
> Nice, but I wish it used space a bit better and needed less clicking. Maybe half the size of icons and use less tabs?

Shrinking the icons makes it look way to blurry as these are large icon files and enlarging them is the only way we can make them look good .Like u said because there is a lot of empty space its part of the reason why i enlarged them to kinda compensate for it 
The tabs have been organized in a such way that each tab corresponds to an rcd_upgrade. So if the upgrade isn't installed its very easy in code to skip all items in that tab else i would have to check against each item if the upgrade allows for that item which is slowler.

Phaseus commented:
Does this also affect borg RCD, which uses energy instead of matter?

SyncIt21 commented:
> Not a fan of this, seems WAY slower than before and harder to navigate with the scrolling Having a big screen in your face doesnt help either

I dont see how this makes navigation harder than the previous interface. For example when selecting an airlock type it was a 3 step operation.
->Select air lock type
->Again open the menu select air lock type
->Select windoor , glass , solid 
->If the air lock type was the last option in the list then you have to click the Next button till you find your airlock type
And finally to change the airlock access you have to again open the menu and select change airlock access which will open up a big window anyway
Another problem is toggling between directional windows & reinforced windows which had just 2 icons and made you check the log to see if you got it correct.
With this interface its as easy as selecting the right tabs & your done. The big window definitly HELPS in making things very clear to the user.

SyncIt21 commented:
> Does this also affect borg RCD, which uses energy instead of matter?

Intresting i  never used that item. Hold on let me check and get back to u. I think it mostly should if it uses an internal RCD but ill make sure

stylemistake commented:
> because there is a lot of empty space its part of the reason why i enlarged them to kinda compensate for it

How about make the window smaller? There's a lot of unused space horizontally, only way to fix this is to make the window smaller.

I also think small icons would be just fine.

Ryusenshu commented:
> > Not a fan of this, seems WAY slower than before and harder to navigate with the scrolling Having a big screen in your face doesnt help either
> 
> I dont see how this makes navigation harder than the previous interface. For example when selecting an airlock type it was a 3 step operation. ->Select air lock type ->Again open the menu select air lock type ->Select windoor , glass , solid ->If the air lock type was the last option in the list then you have to click the Next button till you find your airlock type And finally to change the airlock access you have to again open the menu and select change airlock access which will open up a big window anyway Another problem is toggling between directional windows & reinforced windows which had just 2 icons and made you check the log to see if you got it correct. With this interface its as easy as selecting the right tabs & your done. The big window definitly HELPS in making things very clear to the user.

Except that the big window is only used IF you want to set access to an airlock, and not for the other things you can do with an RCD.
The Window Thing could be fixed with an icon change.
And i have no idea why we even have colored airlocks in RCD anyway, should just be white and glass white
We have an item called airlock painter afterall

So no, another window my controls could get stuck on or can be affected by TGUI crashes, with the only plus of "its all in one window", isnt a good thing

SyncIt21 commented:
> Does this also affect borg RCD, which uses energy instead of matter?

Yup it does . Dosent affect its functioning but uses this interface

SyncIt21 commented:
> > because there is a lot of empty space its part of the reason why i enlarged them to kinda compensate for it
> 
> How about make the window smaller? There's a lot of unused space horizontally, only way to fix this is to make the window smaller.
> 
> I also think small icons would be just fine.

Sure. I can make the icons smaller , ill post a pick of it when im done. But i cant reduce window size unfortunatly because the airlock access panel requires that space to fit perfectly in the window. I dont know how to toggle window sizes based on the selected category or if that is even a thing

MTandi commented:
> I dont see how this makes navigation harder than the previous interface. For example when selecting an airlock type it was a 4 step operation. ->Select air lock mode ->Again open the menu select air lock type ->Select windoor , glass , solid ->If the air lock type was the last option in the list then you have to click the Next button till you find your airlock type And finally to change the airlock access you have to again open the menu and select change airlock access which will open up a big window anyway Another problem is toggling between directional windows & reinforced windows which had just 2 icons and made you check the log to see if you got it correct. With this interface its as easy as selecting the right tabs & your done. The big window definitly HELPS in making things very clear to the user.

You're forgetting about the cursor traveling path. With radial menu you could navigate a lot quicker, especially if you've been working with RCD for a while and know where to click before the next button even appeared. I honestly prefer the radial menu more.

Ryusenshu commented:
> 
> You're forgetting about the cursor traveling path. With radial menu you could navigate a lot quicker, especially if you've been working with RCD for a while and know where to click before the next button even appeared. I honestly prefer the radial menu more.
> 
> And because the window is in the way, you will always want to open and close it. With radial menu, you just hit Z, do a couple clicks and it closes automatically.

As a engi borg main i tend to reset back to "Wall" after every use while on the way to the next destination

MTandi commented:
Please consider making this window optional. Make it open on right clicking RCD, like gas analyzer, and leave the radial menu for those who are used to it.

SyncIt21 commented:
> Please consider making this window optional. Make it open on right clicking RCD, like gas analyzer, and leave the radial menu for those who are used to it.

Sure ill do that then. I wanted to make the RCD more beginner friendly but for the experienced pros they can have their radial menus.

SyncIt21 committed:
Removed Icon Scaling

Can't make them any smaller else they wont become visible

jlsnow301 commented on tgui/packages/tgui/interfaces/RapidConstructionDevice.tsx:
```diff
@@ -0,0 +1,179 @@
+import { AirLockMainSection } from './AirlockElectronics';
+
+type Data = {
+  matterLeft: Number;
+  space: Boolean;
+  silo_upgraded: Boolean;
+  silo_enabled: Boolean;
```
```suggestion
  matterLeft: number;
  space: BooleanLike;
  silo_upgraded: BooleanLike;
  silo_enabled: BooleanLike;
```
same for the other instances

jlsnow301 commented on tgui/packages/tgui/interfaces/RapidConstructionDevice.tsx:
```diff
@@ -0,0 +1,179 @@
+      </Tabs>
+      {shownCategory?.designs.map((design) => (
+        <Button
+          key={design.design_id}
```
will `className="candystripe"` work here? I think it will make it look a little more list-y

stylemistake replied:
Not with buttons, I think.

SyncIt21 replied:
Alternate buttons are colored with stripes. Also the highlighting of the button disappears from alternate options if i move my cursor away from it & only the text remains highlighted

https://user-images.githubusercontent.com/110812394/205607450-cf815ef8-8455-4e50-ae8d-996ff28b07a4.mp4

SyncIt21 replied:
@stylemistake is the icon size alright? I coudnt find better icons for the directional window types as i just reused the radial icons for that if that isnt a problem?

SyncIt21 replied:
> > @Ryusenshu , @MTandi the radials are back check the video above. Any other problems/suggestions?
> 
> Yeah, sure. Just make sure to add that you can right click it to description. Beginners may never try right click to see the beginner friendly window. I noticed the Analyzer window by accident.

Got it

stylemistake replied:
I like the size of directional icons, all icons should be of that size in my opinion, and tighten up the space on buttons. Reusing the icons isn't a problem.

SyncIt21 replied:
I think i just might be able to tighten the spacing after all. Please hold

SyncIt21 replied:
Alright i have done it!!. Button spacing is at the bare minimum without clipping. The scroll bar for the glass air locks is gone!!. I cant reduce it any more than this. Is this alright? @stylemistake 

https://user-images.githubusercontent.com/110812394/205677275-ec315e6f-e0c2-490b-80d1-0d15bc54716b.mp4

SyncIt21 committed:
Update tgui/packages/tgui/interfaces/RapidConstructionDevice.tsx

Proper types

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

SyncIt21 committed:
Radial & TGUI co-exist

Radial Menu Left Click
TGUI Right Click

SyncIt21 committed:
Merge branch 'master' into RCD-TGUI

MTandi commented:
> @Ryusenshu , @MTandi the radials are back check the video above. Any other problems/suggestions?

Yeah, sure. Just make sure to add that you can right click it to description. 
Beginners may never try right click to see the beginner friendly window. I noticed the Analyzer window by accident.

SyncIt21 committed:
Add description

So they know

SyncIt21 commented:
> > @Ryusenshu , @MTandi the radials are back check the video above. Any other problems/suggestions?
> 
> Yeah, sure. Just make sure to add that you can right click it to description. Beginners may never try right click to see the beginner friendly window. I noticed the Analyzer window by accident.

Done

SyncIt21 committed:
Merge branch 'master' into RCD-TGUI

Ryusenshu commented:
Looks good for me

zxaber commented:
Wasn't right-click with the RCD how you deconstructed? How is deconstruct handled now?

To be honest, we should just pick a style and go with it.

MTandi commented:
> Wasn't right-click with the RCD how you deconstructed? How is deconstruct handled now?
> 
> To be honest, we should just pick a style and go with it.

Right click on the RCD itself, not on the tile when you hold it.

SyncIt21 committed:
UI tweaks

1. Bare minimum button spacing
2. Uniform icon sizes

SyncIt21 committed:
Merge branch 'master' into RCD-TGUI

stylemistake commented on code/game/objects/items/RCD.dm:
```diff
@@ -710,7 +711,10 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
-			designs += list(list("design_id" = i, "selected" = ((design["title"] == design_title) && (sub_category == design_category)), "title" = design["title"], "icon" = icon))
+			transform = "scale(1.0)"
+			if(icon == "window0" || icon == "rwindow0")
+				transform = "scale(0.7)"
```
Transform is better to be a function in js to reduce the amount of data that is sent to the client.

stylemistake replied:
I believe `selected` can also be computed client-side, also to reduce the amount of data sent.

SyncIt21 replied:
Ill make it a constant expression in the tsx file for this check

SyncIt21 replied:
Done

SyncIt21 committed:
Merge branch 'master' into RCD-TGUI

SyncIt21 committed:
Merge branch 'master' into RCD-TGUI

SyncIt21 committed:
UI optimizations

1. transform & selected are infered in the client side for RCD & plumbing
2. Removed Stack & nbsp for both plumbing & RCD

SyncIt21 committed:
Merge branch 'master' into RCD-TGUI

SyncIt21 commented:
@stylemistake , @jlsnow301 anything else?

LemonInTheDark commented:
> Please consider making this window optional. Make it open on right clicking RCD, like gas analyzer, and leave the radial menu for those who are used to it.

This is cringe. If it's better we should do it, if it isn't we shouldn't.
Adding changes like this just means we're gonna have to maintain radial rcds in 2 years when none uses the radial anymore.

Mothblocks reviewed (changes requested):
do not make it optional. theres absolutely zero point

Mothblocks commented on code/game/machinery/doors/airlock_electronics.dm:
```diff
@@ -53,9 +53,10 @@
-	. = ..()
-	if(.)
-		return
+	if(!params["rcd"]) //if this wasen't call by the rcd then do these checks else skip them
```
i do not understand this. why is this here?

SyncIt21 replied:
An optimization when the RCD ui_act calls this method. RCD
```
code/game/objects/items/RCD.dm
775    else
776		params["rcd"] = TRUE
777		airlock_electronics.ui_act(action, params)
```
 There is no need for `airlock_electronics` `ui_act()` to call its super method and waste processing time as it's of no use if the RCD calls this method. We can just skip these steps

Mothblocks replied:
This isn't worth it, the cost of a proc call in a function this cold is way too cheap to have this burden. Plus it means a hacker can send the "rcd" parameter and never have to worry about permission checks

SyncIt21 replied:
U right I'll remove it

SyncIt21 replied:
There is a problem tho.  Like u said the permission checks will do a early return if I manually call it from rcd ui_act(). 
I can't move it inside a separate proc either cause again a hacker can just call that proc

SyncIt21 replied:
I don't want to manually copy all the stuff inside ui act and paste it here either cause that would be a big mess

SyncIt21 replied:
I have no solution for this.  Could you suggest something?

SyncIt21 replied:
I'm just gonna move it inside its own proc.  Don't see any other way

SyncIt21 replied:
Now that I think of it don't know why a hacker would call this proc manually when he can just use the interface.  I made a dumb argument nevermind

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -166,6 +173,37 @@ RLD
+	data["matterLeft"] = total_matter
+
+	//silo details
+	data["silo_upgraded"] = upgrade & RCD_UPGRADE_SILO_LINK
```
```suggestion
	data["silo_upgraded"] = !!(upgrade & RCD_UPGRADE_SILO_LINK)
```

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -233,6 +271,94 @@ RLD
+				list("construction_mode" = RCD_FLOORWALL, "icon" = "wallfloor", "title" = "Wall/Floor"),
+				list("construction_mode" = RCD_WINDOWGRILLE, "window_type" = /obj/structure/window, "window_glass" = RCD_WINDOW_NORMAL, "window_size" =  RCD_WINDOW_DIRECTIONAL, "icon" = "windowsize", "title" = "Directional Window"),
+				list("construction_mode" = RCD_WINDOWGRILLE, "window_type" = /obj/structure/window/reinforced, "window_glass" = RCD_WINDOW_REINFORCED, "window_size" =  RCD_WINDOW_DIRECTIONAL, "icon" = "windowtype", "title" = "Directional Reinforced Window"),
+				list("construction_mode" = RCD_WINDOWGRILLE, "window_type" = /obj/structure/window/fulltile, "window_glass" = RCD_WINDOW_NORMAL, "window_size" =  RCD_WINDOW_FULLTILE, "icon" = "window0", "title" = "Full Tile Window"),
```
Make all these strings defines

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -233,6 +271,94 @@ RLD
+	var/design_category = "Structures"
+	var/root_category = "Construction"
+	var/closed = FALSE
+	var/ui_always_active = FALSE ///used by construction_console
```
`///` has to be before the var

SyncIt21 replied:
Yup

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -233,6 +271,94 @@ RLD
+
+	var/static/list/radial_menu = null
+
+	var/design_title = "Wall/Floor" ///english name for the design to check if it was selected or not
```
Same

SyncIt21 replied:
Got it

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -276,6 +402,123 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+		return
+
+	var/list/icon_list = list()
+	var/list/essentials = list(
```
All of this:

1. Needs to be in its own proc
2. Needs to be done only once
3. Should only be done when necessary.

This is all extremely heavy to compute.

SyncIt21 replied:
Why create a proc and have extra method calling overhead when that proc would only be called by Initialize() anyway?. Also this stuff is only invoked once. See the if statement above

```
//Create the radial menu as a static list only at the beginning of the map for all RCD instances to share
	if(!mapload)
		return
```

mapload will be false for all instances of RCD's created after the map has loaded and i store it in a static list so memory isnt a issue either

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -276,6 +402,123 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+
+	var/list/icon_list = list()
+	var/list/essentials = list(
+		'icons/hud/radial.dmi' = list("wallfloor", "windowsize", "windowtype", "grillewindow", "cnorth", "csouth", "ceast", "cwest", "chair", "stool", "windoor", "secure_windoor", "airlock", "airlocktype", "access", "computer_dir"),
```
This all looks like a nightmare to maintain, please make a unit test that ensures all these exist

SyncIt21 replied:
I have no idea how to write unit tests. Also why should i be the only one to do that? The people who wrote the radial menu before me didn't write these checks ,I am using the exact icons as the previous radial menu you can check the video above.
Also this is radial menu stuff so im still waiting on your word to remove this.

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -276,6 +402,123 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+				icon_list["airlock-glass"] = icon
+			else
+				icon = icon(icon = icon_file, icon_state = icon_state)
+				if(icon_state == "window0" || icon_state == "rwindow0")
```
All of this is so hackyyyy, document why you're doing any of this

SyncIt21 replied:
This is "window0"
![Screenshot (57)](https://user-images.githubusercontent.com/110812394/205889462-3992f363-ac55-4a17-88df-48b595794b98.png)

This is grill
![Screenshot (58)](https://user-images.githubusercontent.com/110812394/205889513-4324b34b-9f87-4bdf-80b5-30455c300998.png)

When i blend them together i get this[scaled as i took screen shot from ingame]
![Screenshot (59)](https://user-images.githubusercontent.com/110812394/205890131-202ab188-434f-4fa4-b8e3-17028ea04a7e.png)

Which is the icon i use for Fulltile window

Similarly, "rwindow0" is for Refinforced Fulltile window

I'll put a comment saying that

Also this code is part of the radial menu. Still waiting on your word to scrap this

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -276,6 +402,123 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+		),
+		"Grills & Windows" = list(
+			"Directional Window" = list("icon" = icon_list["windowsize"], "coordinates" = list("Root" = "Construction", "Category" = "Structures", "Index" = 2)),
+			"Directional Reinforced Window" = list("icon" = icon_list["windowtype"], "coordinates" = list("Root" = "Construction", "Category" = "Structures", "Index" = 3)),
```
All these string keys should be defines

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -276,6 +402,123 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+			"Furnishing" = list("icon" = icon_list["chair"], "requires" = RCD_UPGRADE_FURNISHING),
+		),
+		"Grills & Windows" = list(
+			"Directional Window" = list("icon" = icon_list["windowsize"], "coordinates" = list("Root" = "Construction", "Category" = "Structures", "Index" = 2)),
```
What is index and why are we typing them all manually? We will have to adjust them all if we want to change the order?

SyncIt21 replied:
These indices refer to the where the design is located in the static list `root_categories`. for example 
`"coordinates" = list("Root" = "Construction", "Category" = "Structures", "Index" = 2))`
actually refer to
`root_categories["Construction"]["Structures"][2] and that would be the design type which i call` ui_act()`. So yes we would have to type them manually but again this is the radial menu so we gonna scrap that?

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -276,6 +402,123 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+		icon.Blend(icon(icon = airlocks[airlock_name], icon_state = "fill_closed", dir = SOUTH), ICON_OVERLAY)
+		radial_menu["Solid"][airlock_name] = list("icon" = icon, "coordinates" = list("Root" = "AirLocks", "Category" = "Solid AirLocks", "Index" = solid_ref++))
+
+		for(var/exclude as anything in exclusion)
+			if(airlock_name == exclude)
+				skip = TRUE
+				break
```
Replace with `airlock_name in exclude`.

SyncIt21 replied:
Yup . Code part of radial menu waiting for your word to scrap all of this

Mothblocks reviewed (changes requested):
Stopping here but this code is *extremely* hacky and flimsy, and will shatter with any refactors to any of these, making it a lot worse than the current menus in terms of maintainability. Not happy with basically all of it

SyncIt21 commented:
@Mothblocks i would like to make everyone happy and keep the radial menu but if you don't agree ill scrape it. Last call 
Scrape radial menu? Yes/No?

SyncIt21 committed:
Update code/game/objects/items/RCD.dm

Spacing

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 committed:
Fix comments take 1

SyncIt21 committed:
Merge branch 'RCD-TGUI' of https://github.com/SyncIt21/The-TG-Station-Fork into RCD-TGUI

SyncIt21 committed:
Fix comments take 2

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -276,6 +402,123 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+
+		icon = icon(icon = airlocks[airlock_name] , icon_state = "closed" , dir = SOUTH)
+		icon.Blend(icon(icon = airlocks[airlock_name], icon_state = "fill_closed", dir = SOUTH), ICON_OVERLAY)
+		radial_menu["Solid"][airlock_name] = list("icon" = icon, "coordinates" = list("Root" = "AirLocks", "Category" = "Solid AirLocks", "Index" = solid_ref++))
```
Yup ill change that

SyncIt21 replied:
Done

SyncIt21 committed:
Refs & Exclusion

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -276,6 +402,123 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+
+		icon = icon(icon = airlocks[airlock_name] , icon_state = "closed" , dir = SOUTH)
+		icon.Blend(icon(icon = airlocks[airlock_name], icon_state = "fill_closed", dir = SOUTH), ICON_OVERLAY)
+		radial_menu["Solid"][airlock_name] = list("icon" = icon, "coordinates" = list("Root" = "AirLocks", "Category" = "Solid AirLocks", "Index" = solid_ref++))
```
See the static list `root_categories` for example root_categories["AirLocks"]["Solid AirLocks"][1] would point to the first solid airlock design in that list. 
As the name implies `"coordinates" ` tells us where to find the information of this design in that list.
Also code part of radial menu waiting for your word to scrap this

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -276,6 +402,123 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+			continue
+
+		icon = icon(airlocks[airlock_name] , "closed" , SOUTH)
+		radial_menu["Glass"][airlock_name] = list("icon" = icon, "coordinates" = list("Root" = "AirLocks", "Category" = "Glass AirLocks", "Index" = glass_ref++))
```
Code part of radial menu. See my above comment . Same explanation

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -670,81 +649,181 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+
+	//merge airlock_electronics ui data with this
+	var/list/airlock_data = airlock_electronics.ui_data(user)
+	for(var/key as anything in airlock_data)
```
Yup

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -670,81 +649,181 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+			continue
+
+		designs = list() //initialize all designs under this category
+		var/list/target_category =  root_categories[root_category][sub_category]
```
My bad. It's a habit

SyncIt21 committed:
Switch & in for loop

SyncIt21 committed:
Merge branch 'master' into RCD-TGUI

SyncIt21 committed:
airlock in exclusion here too

SyncIt21 committed:
Merge branch 'RCD-TGUI' of https://github.com/SyncIt21/The-TG-Station-Fork into RCD-TGUI

SyncIt21 committed:
Update code/game/objects/items/RCD.dm

Space

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -670,81 +649,181 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+				icon = "[design["title"]]-Glass"
+			else
+				icon = design["icon"]
+			icon = replacetext(icon," ","-") //css class names cannot have spaces in them
```
Thats convenient. thanks

SyncIt21 replied:
Actually i loaded the icon files with `'-'` in the name. This dosen't need to change

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -670,81 +649,181 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+	data["root_categories"] += "AirLock Access"
+
+	//we display the normal tabs only if the selected catagory isnt AirLock Access else we display the AirLockInterface
+	data["display_tabs"] = (root_category != "AirLock Access")
```
Never thought of that. Good point ill do that

SyncIt21 commented on code/game/objects/items/RCD.dm:
```diff
@@ -670,81 +649,181 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+	//main categories
+	data["selected_root"] = root_category
+	data["root_categories"] = list()
+	for(var/category as anything in root_categories)
```
Got it

SyncIt21 commented:
> do not make it optional. theres absolutely zero point

All right. I wanted to wait for a second confirmation but i guess i won't get a different answer. I'll remove the radial menu.

SyncIt21 commented:
@Mothblocks i do agree with you anyway. Its the very reason i made this PR in the first place but since some people became nostalgic about the old radial menu i though i could make them happy as well.

Please consider making this window optional. Make it open on right clicking RCD, like gas analyzer, and leave the radial menu for those who are used to it.

> This is cringe. If it's better we should do it, if it isn't we shouldn't.
> Adding changes like this just means we're gonna have to maintain radial rcds in 2 years when none uses the radial anymore.

But @LemonInTheDark  makes a good point TGUI although error prone & crashing for now will get fixed & become more prominent in the coming days as its way more clean and informative. 

@MTandi , @Ryusenshu i'm sorry bu the maintainers have spoken. The radial menu has to go

SyncIt21 committed:
I forgot basics

1. Fixed if & switch statement
2. Fixed airlock_name put in correct place
3. Moved stuff to client side

SyncIt21 committed:
Merge branch 'RCD-TGUI' of https://github.com/SyncIt21/The-TG-Station-Fork into RCD-TGUI

SyncIt21 committed:
Good Bye Radial Menu

It was nice knowing you :(

SyncIt21 committed:
Ok now good bye for real

SyncIt21 commented:
@Mothblocks you can forgot all comments where i mentioned the radial menu. All that hacky code is gone now. Anything else?

SyncIt21 commented:
Video with all changes implemented is shown above.

SyncIt21 committed:
Merge branch 'master' into RCD-TGUI

SyncIt21 committed:
Merge branch 'tgstation:master' into RCD-TGUI

SyncIt21 committed:
Merge branch 'master' into RCD-TGUI

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -166,6 +173,37 @@ RLD
+				to_chat(usr, span_alert("No silo link detected. Connect to silo via multitool."))
+				return FALSE
+			silo_link = !silo_link
+			to_chat(usr, span_notice("You change \the [src]'s storage link state: [silo_link ? "ON" : "OFF"]."))
```
```suggestion
			to_chat(usr, span_notice("You change [src]'s storage link state: [silo_link ? "ON" : "OFF"]."))
```

the is implicit

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -166,6 +173,37 @@ RLD
+			silo_link = !silo_link
+			to_chat(usr, span_notice("You change \the [src]'s storage link state: [silo_link ? "ON" : "OFF"]."))
+		else
+			to_chat(usr, span_warning("\the [src] doesn't have remote storage connection."))
```
```suggestion
			to_chat(usr, span_warning("[src] doesn't have remote storage connection."))
```

the is implicit

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -670,81 +533,136 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+				icon = "[design["title"]]-Glass"
+			else
+				icon = design["icon"]
+			icon = replacetext(icon," ","-") //css class names cannot have spaces in them
```
I do not understand why you don't want this to be `sanitize_css_class_name`??

SyncIt21 replied:
Done. I did use it but it didn't work, then i realized i had to do the same in the asset file

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -670,81 +533,136 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+/obj/item/construction/rcd/ui_data(mob/user)
+	var/list/data = ..(user)
+
+	data["space"] = TRUE
```
This shouldn't be done through server side. Why is it like this?

SyncIt21 replied:
Just learnt how to use props in typescript.  I'll change it noow

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -670,81 +533,136 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+			if(sub_category == "Solid AirLocks")
+				icon = "[design["title"]]"
+			else if(sub_category == "Glass AirLocks")
+				icon = "[design["title"]]-Glass"
```
This is still way too hacky, if we change glass airlock icon state names then this will go completely undetected.

SyncIt21 replied:
I need to rethink this

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -670,81 +533,136 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+			var/list/design = target_category[i]
+
+			//get/infer icon_state based on category
+			var/icon
```
This should have its name changed to `icon_state` or something because "icon" heavily implies a BYOND icon object.

SyncIt21 replied:
I'll do that

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -670,81 +533,136 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+	switch(action)
+		if("root_category")
+			var/new_root = params["root_category"]
+			if(root_categories[new_root] != null || new_root == "AirLock Access") //is a valid category
```
Why are we referencing specific roots here? What is this for?

SyncIt21 replied:
Airlock access is an special exception because it does not have items to display but it shows the Airlock electronics access panel UI.
So I need an exceptional check for this because it's not defined in the static list root_categories

SyncIt21 replied:
Never mind fixed this

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -670,81 +533,136 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+				furnish_cost = design["furnish_cost"]
+				furnish_delay = design["furnish_delay"]
+
+			if(root_category == "AirLocks")
```
Also why are these all "AirLocks" and not "Airlocks"?

SyncIt21 replied:
I'll change that

Mothblocks commented on code/game/objects/items/RCD.dm:
```diff
@@ -934,32 +851,60 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
+			if(new_choice == null)
+				return
+
+			var/icon/I
```
No single letter variables full `icon`

SyncIt21 replied:
Yup

Mothblocks reviewed (changes requested):
You NEED to change all the strings you're repeating over and over again to defines. It is going to be a nightmare if it's ever changed. They don't need to be global but they need to be typo-protected with defines.

I still *really* do not like how much category specific and icon specific stuff is in here. This entire thing is still extremely flaky

SyncIt21 committed:
Proper grammer

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 committed:
Grammer

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 commented on code/modules/asset_cache/assets/rcd.dm:
```diff
@@ -0,0 +1,62 @@
+
+
+
+
```
Oofff. I'll remove them

SyncIt21 committed:
Merge branch 'master' into RCD-TGUI

SyncIt21 committed:
All about spaces

1. Moved space to client side
2. removed unused space

SyncIt21 committed:
DEFINE Refractoring

1. Created icon flags to customize icons
2. used sanitize_css_name
3. Created defines for variable strings

SyncIt21 committed:
Airlock lower case "l"

1. corrected spelling
2. mered with root_categories list

SyncIt21 committed:
variable refractoring

1. I renamed to icon and moved inside for loop
2. removed as anything

SyncIt21 committed:
Merge branch 'master' into RCD-TGUI

SyncIt21 committed:
Only tab no space

SyncIt21 commented:
@Mothblocks anything else?

stylemistake commented:
You can dynamically adjust width/height of a Window by setting a `key` property on it, and make the key consist of the variables that are expected to resize the window, smth like this below.

```
key={`${selected_tab}`}
width={selected_tab === AIRLOCK_ACCESS ? 456 : 123}
```

SyncIt21 commented:
> You can dynamically adjust width/height of a Window by setting a `key` property on it, and make the key consist of the variables on which window size depends, smth like this below.
> 
> ```
> key={`${selected_tab}`}
> width={selected_tab === AIRLOCK_ACCESS ? 456 : 123}
> ```

Airlock door types also require a height of 456 not just airlock access. Also if we were to add new categories in the futurewe would need to compute new heights for those categories as well. I feel like it woudn't be maintainable to hard code these values

stylemistake commented:
> Airlock door types also require a height of 456 not just airlock access.

Then just have a const that holds an expression that checks whether it's one of those screens, and then use that in the ternary expression.

> Also if we were to add new categories in the future we would need to compute new heights for those categories as well. I feel like it woudn't be maintainable to hard code these values

It would require resizing other UIs that use airlock access/type components, so not a maintainability issue.

SyncIt21 commented:
> > Airlock door types also require a height of 456 not just airlock access.
> 
> Then just have a const that holds an expression that checks whether it's one of those screens, and then use that in the ternary expression. Not sure why scroll is bad, though.
> 
> > Also if we were to add new categories in the future we would need to compute new heights for those categories as well. I feel like it woudn't be maintainable to hard code these values
> 
> New categories would require resizing other UIs that use airlock access/type components, so it's not a maintainability issue, these hardcoded values would be equivalent of just another UI.
> 
> I want core RCD functionality to be as compact as possible.

So i tried this for an hour and it didn't work for a number of reasons

**Reason 1**
When switching between categories we use only the local state variable to keep track of which tab is displayed. This is so much faster
 ```
**const [categoryName, setCategoryName] = useLocalState(
    context,
    'categoryName',
    selected_category
  );**
```
If we want to access this categoryName so we can resize our window based on it we now have to make this global so i did this
```
const CategoryData = {
 selected_category : '';
}
const setCategory = (name) =>{
   CategoryData.selected_category = name
}
```
While this worked it for some reason caused severe lag between switching abs. I think its because the UI is being reloaded whenever i call this function cause its not a local variable so i scraped this idea.

Its either this or we have to pass this to the backend via `act()` and this will just cause overhead 

**Reason 2** 
When switching between the main roots "Construction" , "Airlocks" & "Airlock Access" i need to resize the window in these events too as they will change the categories available. 
For this too be feasible i have to overwrite the users currently selected category when main roots are switched so that the UI can use this newly selected category to determine the new size of the window & hence overwrite the users currently selected choice & category because now those will become invalid. Not only will the users previously selected choice be lost but that would mean invoking `ui_act()` function to change the design & thats way too much overhead

**Reason 3**
I dont simply see a need to make this UI compact because RPD also wastes a lot of space for categories which have a small number of items and it dosen't resize its window. 
![Screenshot (62)](https://user-images.githubusercontent.com/110812394/206692445-41c84cfe-3a60-46e6-b1eb-89bc8aa93eb8.png)
![Screenshot (63)](https://user-images.githubusercontent.com/110812394/206692471-4ce83c93-b095-42da-a324-4cedf9e4fcf6.png)
![Screenshot (64)](https://user-images.githubusercontent.com/110812394/206692491-be95a2af-4ede-4ccf-929e-edede6109de7.png)

Putting this much work here would also mean i need to resize that as well to make these changes consistent in the game & i don't see much benifits from it. 
Everyone's eyes look from top to down in the screen anyway so the extra wasted space in the bottom won't bother anyone especially since everyone is quite used to the RPD interface already. We could use that space to add items under that category in the future so it should be seen as a bonus in my opinion

stylemistake commented:
> Reason 1

You can just reuse `useLocalState` with the same key and default value, and you'll effectively gain access to the same state.

SyncIt21 commented:
> > Reason 1
> 
> You can just reuse `useLocalState` with the same key and default value, and you'll effectively gain access to the same state.

Made this into its own global function
```
const setCategoryName = ([context, selected_category]) => {
  useLocalState(context, 'categoryName', selected_category);
};
```
But now 'categoryName' has become inaccessible
```
const shownCategory =
    categories.find((category) => category.cat_name === categoryName) || ///categoryName syntax error unable to find variable
    categories[0];
```

I have no idea how to do this could you give an example?

SyncIt21 commented:
> > Reason 1
> 
> You can just reuse `useLocalState` with the same key and default value, and you'll effectively gain access to the same state.
> 
> > I dont simply see a need to make this UI compact because RPD also wastes a lot of space for categories which have a small number of items and it dosen't resize its window.
> 
> It used to be a lot more compact, but other guys added more things to it without any consideration 😆, and now it is what it is.

When those ppl clean up their mess ill clean up mine 😄 . Until then i'm too lazy :)

SyncIt21 committed:
Removed unused stuff

TGUI crash actually so this could have been bad

SyncIt21 committed:
Merge branch 'master' into RCD-TGUI

SyncIt21 commented:
Any maintainer interested in suggesting changes or merging this? I have no pending work on this

jlsnow301 commented:
Please tend to mothblock's comment with the ui_act portion, then this can be merged, I have no other reservations against it

SyncIt21 commented:
> > Please tend to mothblock's comment with the ui_act portion, then this can be merged, I have no other reservations against it
> 
> But I did do that right? I mean he asked me to
> 
> * make all strings defines
> * create icons without checking for specific categories,  I did that by making global icon flags which you can specify directly in the design
> * I removed the explicit check for airlock root category
>   is there anything I missed?

Never mind I found it. I'll address the issue

SyncIt21 committed:
airlock ui_act() functionality now inside do_action()

SyncIt21 commented:
@jlsnow301 Done :)

SyncIt21 committed:
Optimization & Bug Fix

1. Only one return statement
2. copied shell value of newly created airlock from rcd airlock

Mothblocks commented:
she

SyncIt21 commented:
> she

Please don't sue me O_O

Mithrun commented:
Needs less tabs please, a lot of empty space and a lot of clicking at the moment..

SyncIt21 commented:
> Needs less tabs please, a lot of empty space and a lot of clicking at the moment..

Each tab corresponds to an RCD upgrade 
Machines tab for RCD Frames upgrade 
Furnishing tab for RCD Furnishing upgrade 
Merging unrelated tabs to reduce their number would be inconsistent and inefficient from a coding perspective 

The extra space is used to  reduce scrolling through airlock door types when you select that tab and since that's the largest section the whole window takes the size of it. Also that extra space can be used to add new designs in the future so it's insurance

Besides the RPD also wastes a lot of lot of space and has a ton of clicking but people got used to that over time and soon this as well

Mothblocks commented:
might look nice as a grid
